src/config.rs

```diff
@@ -190,30 +190,83 @@ impl LayeredConfigs {
+pub enum ConfigPath {
+    /// Path to an existing config file. Note that this is not necessarily
+    /// checked.
+    Existing(PathBuf),
+    /// Path to a new config file in an existing directory.
+    New(PathBuf),
+    /// The path does not exist and the file cannot be created.
+    Unavailable,
+}
+
+impl ConfigPath {
+    fn new(path: PathBuf) -> Self {
+        if path.exists() {
+            return ConfigPath::Existing(path);
+        }
+        if let Some(dir) = path.parent() {
+            if dir.exists() {
+                return ConfigPath::New(path);
+            }
+        }
+        ConfigPath::Unavailable
+    }
+}
+
+impl From<Option<PathBuf>> for ConfigPath {
+    fn from(path: Option<PathBuf>) -> Self {
+        match path {
+            Some(path) => Self::new(path),
+            None => Self::Unavailable,
+        }
+    }
+}
+
+fn config_path() -> Result<ConfigPath, ConfigError> {
     if let Ok(config_path) = env::var("JJ_CONFIG") {
         // TODO: We should probably support colon-separated (std::env::split_paths)
         // paths here
-        Ok(Some(PathBuf::from(config_path)))
-    } else {
-        // TODO: Should we drop the final `/config.toml` and read all files in the
-        // directory?
-        let platform_specific_config_path = dirs::config_dir()
-            .map(|config_dir| config_dir.join("jj").join("config.toml"))
-            .filter(|path| path.exists());
-        let home_config_path = dirs::home_dir()
-            .map(|home_dir| home_dir.join(".jjconfig.toml"))
-            .filter(|path| path.exists());
-        match (&platform_specific_config_path, &home_config_path) {
-            (Some(xdg_config_path), Some(home_config_path)) => Err(ConfigError::AmbiguousSource(
-                xdg_config_path.clone(),
-                home_config_path.clone(),
-            )),
-            _ => Ok(platform_specific_config_path.or(home_config_path)),
-        }
+        return Ok(ConfigPath::Existing(PathBuf::from(config_path)));
+    }
+    // TODO: Should we drop the final `/config.toml` and read all files in the
+    // directory?
+    let xdg_config_path = ConfigPath::from(dirs::config_dir().map(|mut config_dir| {
+        config_dir.push("jj");
+        config_dir.push("config.toml");
+        config_dir
+    }));
+    let home_config_path = ConfigPath::from(dirs::home_dir().map(|mut home_dir| {
+        home_dir.push(".jjconfig.toml");
+        home_dir
+    }));
+    use ConfigPath::*;
+    match (xdg_config_path, home_config_path) {
+        (Existing(xdg_config_path), Existing(home_config_path)) => Err(
+            ConfigError::AmbiguousSource(xdg_config_path, home_config_path),
+        ),
+        (Existing(path), _) | (_, Existing(path)) => Ok(Existing(path)),
+        // Order matters: prefer creating new configs in the platform-specific directory instead of
+        // the home directory.
+        (New(path), _) | (_, New(path)) => Ok(New(path)),
```

Comment:
This does not work because it's unlikely that a `jj` folder exists in the platform dir, which is checked in the `ConfigPath::new`

Reply:
True, but I'm not sure if that means we should do something else.

If the folder exists, creating the config there seems like the right thing. If it doesn't, creating the jj subfolder (providing the platform-specific dir exists) seems questionable. Most tools don't behave this way, and while I could see `jj config set` doing that, `edit` would be more of an issue - we'd have to create the dir before opening the editor, which would leave it empty if the user changed their mind.

Basically I don't think it makes much sense to remove the first pattern and relax (Unavailable, Unavailable) into (_, Unavailable) just because it's very unlikely it's going to make a difference - and I'm not sure there's much of an alternative approach.

Should we just create the jj dir?

Reply:
> Should we just create the jj dir?

If we prefer platform-specific dir (and if the base dir exists), yes, the `jj` subdirectory should be created. 
I think `jj config edit` can also write a sample config file (with comment to the documentation.)

Reply:
The config file in home dir is `.jjconfig.toml` (ew), the config file in platform-specific dir is `jj/config.toml`, so I see no contradiction in creating the `jj` folder, it's just that the current logic doesn't work like that.

For "edit" I would expect it to write a default config there and then edit that, yes

Reply:
OK then, made it create the needed directories and files. Initial config template, if any, can be tackled in a separate PR.

Also changed the code quite a bit to add unit tests.